kv/kvserver/kvflowcontrol/replica_rac2: TestProcessorBasic failed

kv/kvserver/kvflowcontrol/replica_rac2.TestProcessorBasic [failed](https://tanzanite.cluster.engflow.com/invocations/default/2bfd2072-6c14-4127-b862-081d33bae5eb?testReportRun=1&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZzZXJ2ZXIva3ZmbG93Y29udHJvbC9yZXBsaWNhX3JhYzI6cmVwbGljYV9yYWMyX3Rlc3Q=) on master @ [ea1413e0cf4f417fefe2d4406f26aa1ddd934baa](https://github.com/cockroachdb/cockroach/commits/ea1413e0cf4f417fefe2d4406f26aa1ddd934baa):

Fatal error:

```
panic: mutex is not write locked [recovered]
	panic: mutex is not write locked [recovered]
	panic: mutex is not write locked [recovered]
	panic: mutex is not write locked

```
Stack: 

```
goroutine 34 [running]:
testing.tRunner.func1.2({0x25e1960, 0x2df5e20})
	GOROOT/src/testing/testing.go:1631 +0x3f7
testing.tRunner.func1()
	GOROOT/src/testing/testing.go:1634 +0x6b6
panic({0x25e1960?, 0x2df5e20?})
	GOROOT/src/runtime/panic.go:770 +0x132
github.com/cockroachdb/cockroach/pkg/util/leaktest.AfterTest.func2()
	github.com/cockroachdb/cockroach/pkg/util/leaktest/leaktest.go:137 +0x4fa
panic({0x25e1960?, 0x2df5e20?})
	GOROOT/src/runtime/panic.go:770 +0x132
github.com/cockroachdb/datadriven.runDirective.func1.1()
	github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:334 +0x154
panic({0x25e1960?, 0x2df5e20?})
	GOROOT/src/runtime/panic.go:770 +0x132
github.com/cockroachdb/cockroach/pkg/util/syncutil.(*RWMutex).AssertHeld(...)
	github.com/cockroachdb/cockroach/pkg/util/syncutil/mutex_sync_race.go:136
github.com/cockroachdb/cockroach/pkg/kv/kvserver/kvflowcontrol/replica_rac2.(*testRaftNode).LogMarkLocked(0xc00034b440)
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/kvflowcontrol/replica_rac2/processor_test.go:121 +0x128
github.com/cockroachdb/cockroach/pkg/kv/kvserver/kvflowcontrol/replica_rac2.(*processorImpl).InitRaftLocked(0xc0003b3b08, {0x2e15f28, 0x50979c0}, {0x2e1a060, 0xc00034b440})
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/kvflowcontrol/replica_rac2/processor.go:531 +0x18a
github.com/cockroachdb/cockroach/pkg/kv/kvserver/kvflowcontrol/replica_rac2.TestProcessorBasic.func4(0xc000480680, 0xc0000cdd68)
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/kvflowcontrol/replica_rac2/processor_test.go:330 +0x6d6
github.com/cockroachdb/datadriven.RunTest.func1({0x2e2f390, 0xc000480680}, 0xc0000cdd68)
	github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:118 +0x59
github.com/cockroachdb/datadriven.runDirective.func1({0x2e2f390, 0xc000480680}, 0xc0000cdd68, 0xc00038a020)
	github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:337 +0x10d
github.com/cockroachdb/datadriven.runDirective({0x2e2f390, 0xc000480680}, 0xc0000cdd40, 0xc00038a020)
	github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:342 +0x90
github.com/cockroachdb/datadriven.runDirectiveOrSubTest({0x2e2f390, 0xc000480680}, 0xc0000cdd40, {0x0, 0x0}, 0xc00038a020)
	github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:211 +0xa5
github.com/cockroachdb/datadriven.runTestInternal({0x2e2f390, 0xc000480680}, {0xc000200b00, 0xf6}, {0x2dfad20, 0xc0000749a0}, 0xc00038a020, 0x0)
	github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:184 +0xe5
github.com/cockroachdb/datadriven.RunTestAny({0x2e2f390, 0xc000480680}, {0xc000200b00, 0xf6}, 0xc00038a020)
	github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:144 +0x3d1
github.com/cockroachdb/datadriven.RunTest(0xc000480680, {0xc000200b00, 0xf6}, 0xc0001c1e60)
	github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:117 +0xbb
github.com/cockroachdb/cockroach/pkg/kv/kvserver/kvflowcontrol/replica_rac2.TestProcessorBasic(0xc000480680)
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/kvflowcontrol/replica_rac2/processor_test.go:315 +0xad5
testing.tRunner(0xc000480680, 0x2a52248)
	GOROOT/src/testing/testing.go:1689 +0x21f
created by testing.(*T).Run in goroutine 1
	GOROOT/src/testing/testing.go:1742 +0x826
```
<details><summary>Log preceding fatal error</summary>
<p>

```
    datadriven.go:144: 
        /var/lib/engflow/worker/work/2/exec/bazel-out/k8-fastbuild/bin/pkg/kv/kvserver/kvflowcontrol/replica_rac2/replica_rac2_test_/replica_rac2_test.runfiles/com_github_cockroachdb_cockroach/pkg/kv/kvserver/kvflowcontrol/replica_rac2/testdata/processor:23:
        admit-for-eval [1 args]
        <no input to command>
        ----
        admitted: false err: <nil>
    datadriven.go:144: 
        /var/lib/engflow/worker/work/2/exec/bazel-out/k8-fastbuild/bin/pkg/kv/kvserver/kvflowcontrol/replica_rac2/replica_rac2_test_/replica_rac2_test.runfiles/com_github_cockroachdb_cockroach/pkg/kv/kvserver/kvflowcontrol/replica_rac2/testdata/processor:28:
        handle-raft-ready-and-admit [0 args]
        <no input to command>
        ----
        HandleRaftReady:
         Replica.RaftMuAssertHeld
        .....
    datadriven.go:144: 
        /var/lib/engflow/worker/work/2/exec/bazel-out/k8-fastbuild/bin/pkg/kv/kvserver/kvflowcontrol/replica_rac2/replica_rac2_test_/replica_rac2_test.runfiles/com_github_cockroachdb_cockroach/pkg/kv/kvserver/kvflowcontrol/replica_rac2/testdata/processor:34:
        reset [0 args]
        <no input to command>
        ----
        n1,s2,r3: replica=5, tenant=4, enabled-level=not-enabled
    datadriven.go:144: 
        /var/lib/engflow/worker/work/2/exec/bazel-out/k8-fastbuild/bin/pkg/kv/kvserver/kvflowcontrol/replica_rac2/replica_rac2_test_/replica_rac2_test.runfiles/com_github_cockroachdb_cockroach/pkg/kv/kvserver/kvflowcontrol/replica_rac2/testdata/processor:38:
        get-enabled-level [0 args]
        <no input to command>
        ----
        enabled-level: not-enabled
    datadriven.go:144: 
        /var/lib/engflow/worker/work/2/exec/bazel-out/k8-fastbuild/bin/pkg/kv/kvserver/kvflowcontrol/replica_rac2/replica_rac2_test_/replica_rac2_test.runfiles/com_github_cockroachdb_cockroach/pkg/kv/kvserver/kvflowcontrol/replica_rac2/testdata/processor:43:
        set-enabled-level [1 args]
        <no input to command>
        ----
         Replica.RaftMuAssertHeld
    datadriven.go:144: 
        /var/lib/engflow/worker/work/2/exec/bazel-out/k8-fastbuild/bin/pkg/kv/kvserver/kvflowcontrol/replica_rac2/replica_rac2_test_/replica_rac2_test.runfiles/com_github_cockroachdb_cockroach/pkg/kv/kvserver/kvflowcontrol/replica_rac2/testdata/processor:47:
        get-enabled-level [0 args]
        <no input to command>
        ----
        enabled-level: v1-encoding
    datadriven.go:333: 
        panic during /var/lib/engflow/worker/work/2/exec/bazel-out/k8-fastbuild/bin/pkg/kv/kvserver/kvflowcontrol/replica_rac2/replica_rac2_test_/replica_rac2_test.runfiles/com_github_cockroachdb_cockroach/pkg/kv/kvserver/kvflowcontrol/replica_rac2/testdata/processor:53:
        
    datadriven.go:334: -- test log scope end --

ERROR: a panic has occurred!
Details cannot be printed yet because we are still unwinding.
Hopefully the test harness prints the panic below, otherwise check the test logs.

test logs left over in: outputs.zip/logTestProcessorBasic2408082302
    datadriven.go:334: panic: mutex is not write locked
--- FAIL: TestProcessorBasic (0.01s)
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=1</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/storage
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestProcessorBasic.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>


Jira issue: CRDB-42297

kv/kvserver/kvflowcontrol/replica_rac2.TestProcessorBasic [failed](https://tanzanite.cluster.engflow.com/invocations/default/ea9d0068-3938-40fe-95c1-1b15bfa05892?testReportRun=1&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZzZXJ2ZXIva3ZmbG93Y29udHJvbC9yZXBsaWNhX3JhYzI6cmVwbGljYV9yYWMyX3Rlc3Q=) on master @ [285460a47358406e65df9ff3deb5a0764f58a2bf](https://github.com/cockroachdb/cockroach/commits/285460a47358406e65df9ff3deb5a0764f58a2bf):

Fatal error:

```
panic: mutex is not write locked [recovered]
	panic: mutex is not write locked [recovered]
	panic: mutex is not write locked [recovered]
	panic: mutex is not write locked

```
Stack: 

```
goroutine 34 [running]:
testing.tRunner.func1.2({0x25e1cc0, 0x2df6440})
	GOROOT/src/testing/testing.go:1631 +0x3f7
testing.tRunner.func1()
	GOROOT/src/testing/testing.go:1634 +0x6b6
panic({0x25e1cc0?, 0x2df6440?})
	GOROOT/src/runtime/panic.go:770 +0x132
github.com/cockroachdb/cockroach/pkg/util/leaktest.AfterTest.func2()
	github.com/cockroachdb/cockroach/pkg/util/leaktest/leaktest.go:137 +0x4fa
panic({0x25e1cc0?, 0x2df6440?})
	GOROOT/src/runtime/panic.go:770 +0x132
github.com/cockroachdb/datadriven.runDirective.func1.1()
	github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:334 +0x154
panic({0x25e1cc0?, 0x2df6440?})
	GOROOT/src/runtime/panic.go:770 +0x132
github.com/cockroachdb/cockroach/pkg/util/syncutil.(*RWMutex).AssertHeld(...)
	github.com/cockroachdb/cockroach/pkg/util/syncutil/mutex_sync_race.go:136
github.com/cockroachdb/cockroach/pkg/kv/kvserver/kvflowcontrol/replica_rac2.(*testRaftNode).LogMarkLocked(0xc00035d400)
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/kvflowcontrol/replica_rac2/processor_test.go:125 +0x128
github.com/cockroachdb/cockroach/pkg/kv/kvserver/kvflowcontrol/replica_rac2.(*processorImpl).InitRaftLocked(0xc00015c6c8, {0x2e16548, 0x50989c0}, {0x2e1a640, 0xc00035d400})
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/kvflowcontrol/replica_rac2/processor.go:536 +0x18a
github.com/cockroachdb/cockroach/pkg/kv/kvserver/kvflowcontrol/replica_rac2.TestProcessorBasic.func4(0xc00049e820, 0xc0000d7f18)
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/kvflowcontrol/replica_rac2/processor_test.go:334 +0x6d6
github.com/cockroachdb/datadriven.RunTest.func1({0x2e2f9f0, 0xc00049e820}, 0xc0000d7f18)
	github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:118 +0x59
github.com/cockroachdb/datadriven.runDirective.func1({0x2e2f9f0, 0xc00049e820}, 0xc0000d7f18, 0xc00039ce60)
	github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:337 +0x10d
github.com/cockroachdb/datadriven.runDirective({0x2e2f9f0, 0xc00049e820}, 0xc0000d7ef0, 0xc00039ce60)
	github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:342 +0x90
github.com/cockroachdb/datadriven.runDirectiveOrSubTest({0x2e2f9f0, 0xc00049e820}, 0xc0000d7ef0, {0x0, 0x0}, 0xc00039ce60)
	github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:211 +0xa5
github.com/cockroachdb/datadriven.runTestInternal({0x2e2f9f0, 0xc00049e820}, {0xc000036c00, 0xf6}, {0x2dfb340, 0xc0000749e8}, 0xc00039ce60, 0x0)
	github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:184 +0xe5
github.com/cockroachdb/datadriven.RunTestAny({0x2e2f9f0, 0xc00049e820}, {0xc000036c00, 0xf6}, 0xc00039ce60)
	github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:144 +0x3d1
github.com/cockroachdb/datadriven.RunTest(0xc00049e820, {0xc000036c00, 0xf6}, 0xc00003a240)
	github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:117 +0xbb
github.com/cockroachdb/cockroach/pkg/kv/kvserver/kvflowcontrol/replica_rac2.TestProcessorBasic(0xc00049e820)
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/kvflowcontrol/replica_rac2/processor_test.go:319 +0xad5
testing.tRunner(0xc00049e820, 0x2a52848)
	GOROOT/src/testing/testing.go:1689 +0x21f
created by testing.(*T).Run in goroutine 1
	GOROOT/src/testing/testing.go:1742 +0x826
```
<details><summary>Log preceding fatal error</summary>
<p>

```
    datadriven.go:144: 
        /var/lib/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/bin/pkg/kv/kvserver/kvflowcontrol/replica_rac2/replica_rac2_test_/replica_rac2_test.runfiles/com_github_cockroachdb_cockroach/pkg/kv/kvserver/kvflowcontrol/replica_rac2/testdata/processor:23:
        admit-for-eval [1 args]
        <no input to command>
        ----
        admitted: false err: <nil>
    datadriven.go:144: 
        /var/lib/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/bin/pkg/kv/kvserver/kvflowcontrol/replica_rac2/replica_rac2_test_/replica_rac2_test.runfiles/com_github_cockroachdb_cockroach/pkg/kv/kvserver/kvflowcontrol/replica_rac2/testdata/processor:28:
        handle-raft-ready-and-admit [0 args]
        <no input to command>
        ----
        HandleRaftReady:
         Replica.RaftMuAssertHeld
        .....
    datadriven.go:144: 
        /var/lib/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/bin/pkg/kv/kvserver/kvflowcontrol/replica_rac2/replica_rac2_test_/replica_rac2_test.runfiles/com_github_cockroachdb_cockroach/pkg/kv/kvserver/kvflowcontrol/replica_rac2/testdata/processor:34:
        reset [0 args]
        <no input to command>
        ----
        n1,s2,r3: replica=5, tenant=4, enabled-level=not-enabled
    datadriven.go:144: 
        /var/lib/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/bin/pkg/kv/kvserver/kvflowcontrol/replica_rac2/replica_rac2_test_/replica_rac2_test.runfiles/com_github_cockroachdb_cockroach/pkg/kv/kvserver/kvflowcontrol/replica_rac2/testdata/processor:38:
        get-enabled-level [0 args]
        <no input to command>
        ----
        enabled-level: not-enabled
    datadriven.go:144: 
        /var/lib/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/bin/pkg/kv/kvserver/kvflowcontrol/replica_rac2/replica_rac2_test_/replica_rac2_test.runfiles/com_github_cockroachdb_cockroach/pkg/kv/kvserver/kvflowcontrol/replica_rac2/testdata/processor:43:
        set-enabled-level [1 args]
        <no input to command>
        ----
         Replica.RaftMuAssertHeld
    datadriven.go:144: 
        /var/lib/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/bin/pkg/kv/kvserver/kvflowcontrol/replica_rac2/replica_rac2_test_/replica_rac2_test.runfiles/com_github_cockroachdb_cockroach/pkg/kv/kvserver/kvflowcontrol/replica_rac2/testdata/processor:47:
        get-enabled-level [0 args]
        <no input to command>
        ----
        enabled-level: v1-encoding
    datadriven.go:333: 
        panic during /var/lib/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/bin/pkg/kv/kvserver/kvflowcontrol/replica_rac2/replica_rac2_test_/replica_rac2_test.runfiles/com_github_cockroachdb_cockroach/pkg/kv/kvserver/kvflowcontrol/replica_rac2/testdata/processor:53:
        
    datadriven.go:334: -- test log scope end --

ERROR: a panic has occurred!
Details cannot be printed yet because we are still unwinding.
Hopefully the test harness prints the panic below, otherwise check the test logs.

test logs left over in: outputs.zip/logTestProcessorBasic456755129
    datadriven.go:334: panic: mutex is not write locked
--- FAIL: TestProcessorBasic (0.01s)
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=1</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestProcessorBasic.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>

--- a/pkg/kv/kvserver/kvflowcontrol/replica_rac2/processor_test.go
+++ b/pkg/kv/kvserver/kvflowcontrol/replica_rac2/processor_test.go
@@ -110,25 +110,25 @@ type testRaftNode struct {
 }
 
 func (rn *testRaftNode) TermLocked() uint64 {
-	rn.r.mu.AssertHeld()
+	rn.r.mu.AssertRHeld()
 	fmt.Fprintf(rn.b, " RaftNode.TermLocked() = %d\n", rn.term)
 	return rn.term
 }
 
 func (rn *testRaftNode) LeaderLocked() roachpb.ReplicaID {
-	rn.r.mu.AssertHeld()
+	rn.r.mu.AssertRHeld()
 	fmt.Fprintf(rn.b, " RaftNode.LeaderLocked() = %s\n", rn.leader)
 	return rn.leader
 }
 
 func (rn *testRaftNode) LogMarkLocked() rac2.LogMark {
-	rn.r.mu.AssertHeld()
+	rn.r.mu.AssertRHeld()
 	fmt.Fprintf(rn.b, " RaftNode.LogMarkLocked() = %+v\n", rn.mark)
 	return rn.mark
 }
 
 func (rn *testRaftNode) NextUnstableIndexLocked() uint64 {
-	rn.r.mu.AssertHeld()
+	rn.r.mu.AssertRHeld()
 	fmt.Fprintf(rn.b, " RaftNode.NextUnstableIndexLocked() = %d\n", rn.nextUnstableIndex)
 	return rn.nextUnstableIndex
 }
@@ -330,8 +330,10 @@ func TestProcessorBasic(t *testing.T) {
 				var mark rac2.LogMark
 				d.ScanArgs(t, "log-term", &mark.Term)
 				d.ScanArgs(t, "log-index", &mark.Index)
+				r.mu.Lock()
 				r.initRaft(mark)
 				p.InitRaftLocked(ctx, r.raftNode)
+				r.mu.Unlock()
 				return builderStr()
 
 			case "set-raft-state":

